File: day04/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)


type Card struct {
    WinningNums map[int]struct{}
    MyNums map[int]struct{}
}

func makeNumberSet(s string) map[int]struct{} {
    nums := strings.Split(s, " ")
    numMap := make(map[int]struct{})
    for _, num := range nums {
        if num == "" {
            continue
        }
        
        num, _ := strconv.Atoi(strings.Trim(num, " "))
        numMap[num] = struct{}{}
    }

    return numMap
}

func readInput(path string) []Card  {
	file, _ := os.Open(path)
	defer file.Close()
	scanner := bufio.NewScanner(file)

    cards := make([]Card, 0)
	for scanner.Scan() {
		line := scanner.Text()
        
        numbers := strings.Split(line, ":")[1]
        groups := strings.Split(numbers, " | ")
        
        cards = append(cards, Card{
            WinningNums: makeNumberSet(groups[0]),
            MyNums: makeNumberSet(groups[1]),
        })
    }
	return cards
}

func part1(cards []Card) {
    total := 0.0
    for _, card := range cards {
        count := 0
        for num := range card.MyNums {
            if _, ok := card.WinningNums[num]; ok {
                count++
            }
        }

        total += math.Pow(2,float64(max(0, count - 1)))  
    }	

    println(int(total))
}

func part2(cards []Card) {
    cardValues := make([]int, len(cards))
    for i := range cardValues {
        cardValues[i] = 1
    }

    for i, card := range cards {
        count := 0
        for num := range card.MyNums {
            if _, ok := card.WinningNums[num]; ok {
                count++
            }
        }

        for k := i + 1; k <= i + count; k++ {
            cardValues[k] += cardValues[i]
        }
    }	

    // Sum the card values
    total := 0
    for _, v := range cardValues {
        total += v
    }

    println(int(total))
}

func main() {
	// Get input
	if len(os.Args) != 3 {
		println(len(os.Args))
		println("Usage: [INPUT FILE] [PART 1 or 2]")
	} else {
		inputFilePath := os.Args[1]
		part := os.Args[2]
		input := readInput(inputFilePath)
        fmt.Println(input)
		if part == "1" {
			part1(input)
		} else {
			part2(input)
		}
	}
}

File: day04/main.py
test = """Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
"""


def num_matching(card_line):
    nums = card_line.split(":")[1]
    winning_nums, my_nums = nums.split("|")
    print(winning_nums, my_nums)

    winning_nums = set(map(int, filter(bool, winning_nums.split(" "))))
    my_nums = set(map(int, filter(bool, my_nums.split(" "))))

    return sum((1 for num in my_nums if num in winning_nums))


def part1(input_: str):
    total = 0
    for line in input_.splitlines(keepends=False):
        if matching := num_matching(line):
            total += (matching - 1) ** 2


def part2(input_: str):
    lines = input_.splitlines(keepends=False)
    n = len(lines)
    counts = [1] * n
    for i, line in enumerate(lines):
        matching = num_matching(line)
        print(i + 1, "matching", matching)
        for j in range(i + 1, i + matching + 1):
            counts[j] += counts[i]

    print(counts)
    print(sum(counts))


with open("input.txt", "r") as f:
    input = f.read()

part1(input)
part2(input)
